src/operator/tensor/control_flow_op.h

```diff
@@ -0,0 +1,596 @@
+ * the condition's elements.
+ * This should only be invoked by req_type=kWriteInplace.
+ */
+template<>
+struct where_batch<kWriteInplace> {
+  // DType is the output data type
+  // CType is the condition data type
+  template<typename DType, typename CType>
+  MSHADOW_XINLINE static void Map(int i, DType* out, const CType* cond,
+                                  const DType* x, const DType* y, int M) {
+    out[i] = (cond[i/M] != 0? x[i] : y[i]);
+  }
+};
+
+/*! \brief Choose elements from x or y depending on condition
+ * The condition is a vector whose size is the same as the
+ * x's first dim size.
+ * The returned array is formed by rows from x or y depending on
+ * the condition's elements.
+ * This should only be invoked by req_type=kAddTo.
+ */
+template<>
+struct where_batch<kAddTo> {
+  // DType is the output data type
+  // CType is the condition data type
+  template<typename DType, typename CType>
+  MSHADOW_XINLINE static void Map(int i, DType* out, const CType* cond,
+                                  const DType* x, const DType* y, int M) {
+    out[i] += (cond[i/M] != 0? x[i] : y[i]);
+  }
+};
+
+/*!
+ * \brief Primary template for calculating grad[x].
+ * Will be specialized for req = kNullOp, kWriteTo, kWriteInplace, kAddTo.
+ */
+template<int req>
+struct where_x_backward;
+
+/*!
+ * \brief Specialized backward template for
+ * calculating grad[x] when req_type = kNullOp
+ */
+template<>
+struct where_x_backward<kNullOp> {
+  // DType is the output data type
+  // CType is condition data type
+  template<typename DType, typename CType>
+  MSHADOW_XINLINE static void Map(int i, DType* grad_x,
+                                  const DType* grad_in,
+                                  const CType* cond) {}
+};
+
+/*!
+ * \brief Specialized backward template for
+ * calculating grad[x] when req_type = kWriteTo
+ */
+template<>
+struct where_x_backward<kWriteTo> {
+  // DType is the output data type
+  // CType is condition data type
+  template<typename DType, typename CType>
+  MSHADOW_XINLINE static void Map(int i, DType* grad_x,
+                                  const DType* grad_in,
+                                  const CType* cond) {
+    grad_x[i] = (0 != cond[i]? grad_in[i] : static_cast<DType>(0));
+  }
+};
+
+/*!
+ * \brief Specialized backward template for
+ * calculating grad[x] when req_type = kWriteInplace
+ */
+template<>
+struct where_x_backward<kWriteInplace> {
+  // DType is the output data type
+  // CType is condition data type
+  template<typename DType, typename CType>
+  MSHADOW_XINLINE static void Map(int i, DType* grad_x,
+                                  const DType* grad_in,
+                                  const CType* cond) {
+    grad_x[i] = (0 != cond[i]? grad_in[i] : static_cast<DType>(0));
+  }
+};
+
+/*!
+ * \brief Specialized backward template for
+ * calculating grad[x] when req_type = kAddTo
+ */
+template<>
+struct where_x_backward<kAddTo> {
```

Comment:
where_x and where_y can be merged into one
template<bool neg>
where {
MAP(...) {
   (0 == cond[i])^neg ? ... : ...
}